modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,425 @@
+		return
+	}
+
+	max := u.maxFn(tenant)
+	if max == 0 {
+		max = u.cfg.MaxCardinality
+	}
+
+	var (
+		now                             = time.Now().Unix()
+		data                            = u.getTenant(tenant)
+		mapping, buffer1, buffer2, last = data.GetBuffersForDimensions(dimensions)
+	)
+
+	for _, batch := range batches {
+		unaccountedForBatchData, totalSpanCount := nonSpanDataLength(batch)
+
+		if totalSpanCount == 0 {
+			// Mainly to prevent a panic below, but is this even possible?
+			continue
+		}
+
+		// This is 1/Nth of the unaccounted for batch data that gets added to each span.
+		// Adding this incrementally as we go through the spans is the fastest method, but
+		// loses some precision. The other (original) implementation is to record span counts
+		// per series into a map and reconcile at the end. That method has more accurate data because
+		// it performs the floating point math once on the total, instead of accumulating 1/N + 1/N ... errors.
+		batchPortion := int(math.RoundToEven(float64(unaccountedForBatchData) / float64(totalSpanCount)))
+
+		// To account for the accumulated error we dump the remaining delta onto the first span, which can be negative.
+		// The result ensures the total recorded bytes matches the input.
+		firstSpanPortion := unaccountedForBatchData - batchPortion*totalSpanCount
+
+		// Reset value buffer for every batch.
+		for k := range buffer1 {
+			buffer1[k] = missingLabel
+		}
+
+		if batch.Resource != nil {
+			for _, a := range batch.Resource.Attributes {
+				v := a.Value.GetStringValue()
+				if v == "" {
+					continue
+				}
+				for _, m := range mapping {
+					if m.from == a.Key {
+						buffer1[m.to] = v
+						break
+					}
+				}
+			}
+		}
+
+		var bucket *bucket
+
+		for i, ss := range batch.ScopeSpans {
+			for j, s := range ss.Spans {
+				sz := s.Size()
+				sz += protoLengthMath(sz)
+				sz += batchPortion // Incrementally add 1/Nth worth of the unaccounted for batch data
+				if i == 0 && j == 0 {
+					sz += firstSpanPortion
+				}
+
+				// Reset to batch values to for some spans having missing values.
+				copy(buffer2, buffer1)
+
+				for _, a := range s.Attributes {
+					v := a.Value.GetStringValue()
+					if v == "" {
+						continue
+					}
+					for _, m := range mapping {
+						if m.from == a.Key {
+							buffer2[m.to] = v
+							break
+						}
+					}
+				}
+
+				// Every span can be a different series.
+				// If the values buffer hasn't changed then we
+				// know it's the same bucket and avoid hashing again.
+				// This shows up in 2 common cases:
+				//  - Dimensions are only resource attributes
+				//  - Runs of spans with the same attributes
+				// NOTE - Not happy about the slices.Equal to detect when
+				// to rehash, but couldn't figure out a better way for now.
+				// The difficulty is tracking bucket dirty status while
+				// resetting to batch values and recording the span values.
+				if bucket == nil || !slices.Equal(buffer2, last) {
```

Comment:
i feel like the normal case here is that people will configure resource level attributes only so we will almost always use the previously grabbed bucket.

the best idea i have to dodge `slices.Equal` is if we configured each dimension with resource or span level. then we would only search those attributes that could match and we'd know if we needed to look in span level at all. 

that might not work with the broader goals of the project though if we intend people to specify attribute names only w/o the overhead of knowing its span or resource.

Reply:
Yes we discussed some of this in: https://github.com/grafana/tempo/pull/4162#discussion_r1793856850  For now I'd like to leave the configuration as-is (scopeless) because of the broader goals of the project and also it's similar to how dimensions for the metrics-generator are configured.